### PR TITLE
Simple draw rectangle implementation for feat #607 https://github.com…

### PR DESCRIPTION
**Pull request by mmathew23**
…/kornia/kornia/issues/607

### Description
As part of feat #607, this is a basic implementation of non differentiable draw rectangle.

### Status
WIP

### Types of changes

- [x] New tests added to cover the changes
- [x] New Feature

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [x] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [x] Update notebooks & documentation if necessary
    -not necessary yet
### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/utils/draw.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+from typing import Optional, Union
+
+import torch
+
+
+def rectangle(
+        image: torch.Tensor,
+        boxes: torch.Tensor,
+        color: Optional[Union[float, torch.Tensor]] = 1.0,
```
Comment (edgarriba) [inline kornia/utils/draw.py:9]: optional it's only used for `None`

File: kornia/utils/draw.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        boxes (torch.Tensor): represents number of rectangles to draw in BxNx4
+            N is the number of boxes to draw per batch
+            4 is in (top_left.y, top_left.x, bot_right.y, bot_right.x) ie h1, w1, h2, w2
+        color (int or torch.Tensor): can be a float or tensor if image channels > 1
+            if float is used and channels > 1, float will be broadcasted to all channels
+        fill: bool is a flag used to fill the boxes with color if True
+
+    Returns:
+        torch.Tensor: This operation modifies image inplace but also returns the drawn tensor for convenience
+    """
```
Comment (edgarriba) [inline kornia/utils/draw.py:24]: end string lines with `.` and provide a short snippet example with the usage

File: kornia/utils/draw.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        image (torch.Tensor): is tensor of BxCxHxW
+        boxes (torch.Tensor): represents number of rectangles to draw in BxNx4
+            N is the number of boxes to draw per batch
+            4 is in (top_left.y, top_left.x, bot_right.y, bot_right.x) ie h1, w1, h2, w2
+        color (int or torch.Tensor): can be a float or tensor if image channels > 1
+            if float is used and channels > 1, float will be broadcasted to all channels
+        fill: bool is a flag used to fill the boxes with color if True
+
+    Returns:
+        torch.Tensor: This operation modifies image inplace but also returns the drawn tensor for convenience
```
Comment (edgarriba) [inline kornia/utils/draw.py:23]: indicate output shape

File: kornia/utils/draw.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    """
+
```
Comment (edgarriba) [inline kornia/utils/draw.py:25]: assert type and shape assertions with meaningful error messages

File: kornia/utils/draw.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        boxes: torch.Tensor,
+        color: Optional[Union[float, torch.Tensor]] = 1.0,
+        fill: Optional[bool] = False) -> torch.Tensor:
+    """Draws N rectangles on a batch of image tensors
+
+    Args:
+        image (torch.Tensor): is tensor of BxCxHxW
+        boxes (torch.Tensor): represents number of rectangles to draw in BxNx4
+            N is the number of boxes to draw per batch
+            4 is in (top_left.y, top_left.x, bot_right.y, bot_right.x) ie h1, w1, h2, w2
```
Comment (edgarriba) [inline kornia/utils/draw.py:17]: make sure we follow the same convention every where in the library. /cc @shijianjian probably we should add somewhere documentation about our convention for bounding boxes.

File: kornia/utils/draw.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    _, num_boxes, num_points = boxes.shape
+
+    # clone boxes, in case it's been expanded assignment from clipping causes problems
+    boxes = boxes.clone()
+
+    # clip boxes to hxw bounds
+    boxes[:, :, ::2] = torch.clamp(boxes[:, :, ::2], 0, h - 1)
+    boxes[:, :, 1::2] = torch.clamp(boxes[:, :, 1::2], 0, w - 1)
+    diff = boxes[:, :, 2:] - boxes[:, :, :2]
+    if torch.any(diff < 0):
```
Comment (edgarriba) [inline kornia/utils/draw.py:36]: wouldn't be this very expensive to compute every time ?

File: kornia/utils/draw.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        raise ValueError("All points must correspond to sequent top left followed by bottom right")
+
+    if type(color) is float:
+        color = torch.tensor([color])
```
Comment (edgarriba) [inline kornia/utils/draw.py:40]: pass the input `device` and `dtype`

File: kornia/utils/draw.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        for n in range(num_boxes):
+            if fill:
+                image[b, :, int(boxes[b, n, 0]):int(boxes[b, n, 2]) + 1,
+                      int(boxes[b, n, 1]):int(boxes[b, n, 3]) + 1] = color[:, None, None]
+            else:
+                image[b, :, int(boxes[b, n, 0]):int(boxes[b, n, 2]) + 1, int(boxes[b, n, 1])] = color[:, None]
+                image[b, :, int(boxes[b, n, 0]):int(boxes[b, n, 2]) + 1, int(boxes[b, n, 3])] = color[:, None]
+                image[b, :, int(boxes[b, n, 0]), int(boxes[b, n, 1]):int(boxes[b, n, 3]) + 1] = color[:, None]
+                image[b, :, int(boxes[b, n, 2]), int(boxes[b, n, 1]):int(boxes[b, n, 3]) + 1] = color[:, None]
+
```
Comment (edgarriba) [inline kornia/utils/draw.py:52]: I believe you can vectorise this with `boxes.long()`. Another thing to take into account is line line thickness (maybe for next PR)

File: kornia/utils/draw.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
+from typing import Optional, Union
+
+import torch
+
+
+def rectangle(
+        image: torch.Tensor,
+        boxes: torch.Tensor,
```
Comment (edgarriba) [inline kornia/utils/draw.py:8]: I'm checking now opencv API and they assume just two points simplifying and leaving to the user to compute top-left and bottom-right. Maybe we should consider this API. See: https://docs.opencv.org/master/d6/d6e/group__imgproc__draw.html#ga07d2f74cadcf8e305e810ce8eed13bc9

They also support passing rectangle, however we live in Python world and as for now we don't have a clear way to overload function signatures. /cc @ducha-aiki Which API should we take here ?

File: test/utils/test_draw.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+import torch
+import kornia.testing as utils
+from kornia.utils import rectangle
+
+
+class TestRectangle:
+    @pytest.mark.parametrize('batch', (4, 17))
+    @pytest.mark.parametrize('color', (1.0, 0.5))
+    def test_smoke(self, device, batch, color):
```
Comment (edgarriba) [inline test/utils/test_draw.py:11]: you need to pass `dtype` as well

File: test/utils/test_draw.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
+import pytest
+
+import torch
+import kornia.testing as utils
+from kornia.utils import rectangle
+
+
+class TestRectangle:
```
Comment (edgarriba) [inline test/utils/test_draw.py:8]: consider inheriting from [`BaseTest`](https://github.com/kornia/kornia/blob/b9ffe7efcba7399daeeb8028f10c22941b55d32d/test/enhance/test_normalize.py#L223)

File: test/utils/test_draw.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    @pytest.mark.parametrize('batch', (4, 17))
+    @pytest.mark.parametrize('color', (1.0, 0.5))
+    def test_smoke(self, device, batch, color):
+        black_image = torch.zeros(batch, 1, 3, 3, device=device)  # 1 channel 3x3 black_image
+        points = torch.tensor([1., 1., 1., 1.]).to(device).expand(batch, 1, 4)  # single pixel rectangle
```
Comment (edgarriba) [inline test/utils/test_draw.py:13]: it's recommended to initialize tensor as `torch.tensor(..., device=device, dtype=dtype)`

**Review by edgarriba — CHANGES_REQUESTED** [review]
Good starting point. Check my comments

File: kornia/utils/draw.py (hunk 1 above)
Comment (mmathew23) [reply kornia/utils/draw.py:9]: Yes, will fix

File: kornia/utils/draw.py (hunk 8 above)
Comment (mmathew23) [reply kornia/utils/draw.py:52]: Noted RE: boxes.long()

Line thickness is straightforward but has an added question of how to connect the rectangle lines at the corners. It looks like opencv uses rounded circles at line ends with radius proportional to line thickness. The exact implementation depends on the line type used, so was thinking to table that decision until we're happy with a basic line and rectangle implementation.

File: kornia/utils/draw.py (hunk 6 above)
Comment (mmathew23) [reply kornia/utils/draw.py:36]: @edgarriba I don't believe torch.any should be very expensive, but maybe you're referring to the strided boxes on line 33 and 34?

File: kornia/utils/draw.py (hunk 9 above)
Comment (mmathew23) [reply kornia/utils/draw.py:8]: @ducha-aiki Any thoughts on what to take? Opencv style is simplified, but I guess you couldn't draw a rotated rectangle unless all 4 points are specified.

File: kornia/utils/draw.py (hunk 6 above)
Comment (edgarriba) [reply kornia/utils/draw.py:36]: I refer to the `torch.any` - We also had a short discussion about this kind of value checks with @shijianjian and @FGeri for the augmentations e.g. to check whether a tensor image is bounded between 0/1 but since we are dealing with batched tensors, it can be a bit tricky to determine which sample within the batch is the affected by that condition. If it's only one it's fine, but in case of having multiple not sure what's the proper approach to follow. This also goes to the direction of tensor oriented programming practices.

File: kornia/utils/draw.py (hunk 9 above)
Comment (edgarriba) [reply kornia/utils/draw.py:8]: This can be a strong argument to follow the approach you propose instead of OpenCV one. The case you mention I believe that can be simply generalized by first having a `draw_line` routine that could take `BxCxHxW` and `BxNx2`. Later I believe that can be extended with `draw_rectangle` for `BxNx4` and finally `draw_polyline` `BxNxN`. BTW, been thinking about naming for the API and I believe that `K.draw_rectangle` is more explicit than just `K.rectangle`. What do you think guys ?

**Comment by edgarriba** [comment]
@mmathew23 any update on this ? We are releasing v0.5 in a week - would be great to include this PR in experimental model :)

**Comment by mmathew23** [comment]
@edgarriba, Apologies I didn't see the comments from last month, will review your responses and let you know.

**Comment by mmathew23** [comment]
I wasn't able to carve out enough time to work on this. I'll make sure to get this in for the next one!

**Comment by edgarriba** [comment]
@mmathew23 you still on time. We encountered few issues - so we might delay a bit the release :)

**Comment by edgarriba** [comment]
@mmathew23 what's the status of this PR - Can be used an experimental feature ? @anguelos @lferraz @ducha-aiki  @shijianjian could you take a look at it and provide feedback ?

**Comment by mmathew23** [comment]
@edgarriba Will start working on this, and submit PR with changes. You can let me know if it needs further refinement at that point. Does that sound ok?

**Comment by edgarriba** [comment]
@shijianjian can you give some thoughts about this API ? You were suggesting some changes on how we represent bboxes.

File: kornia/utils/draw.py (hunk 5 above)
Comment (shijianjian) [reply kornia/utils/draw.py:17]: I was thinking either to use [y, x, h, w] or [x1, y1, x2, y2] to represent the bounding boxes, for a more meaningful gradient. Most bounding box algorithms are using the conversion of [y, x, h, w] since it more close to a regression problem than outputting two points. That is why I think we might follow the same conversion as [y, x, h, w].

**Comment by anguelos** [comment]
I favor a low level discrete implementation and optionally a higher level version including sub-pixel precision and rendering realized with anti-aliasing.

In what concerns the discrete location point:
I have to say I am in favor of [inclusive-left, inclusive-top, exclusive-right, exclusive-bottom]. 
I am against [Batch x (#Rects per image) x 4] as it arbitrary assumes a fixed amount of rects per image and introduces  the need for an invalid rect.
I suggest [(#Rects per batch) x 5] where 5 is [ImageInBatch, left, top, right, bottom]

Instead of an optional dimension for rendering channels, we should have an optional "color" parameter being represented by a tensor [(#Rects per batch) x #channels] and an optional ignore_channel boolean tensor of the size [(#Rects per batch) x #channels]. Both these tensors could be optional and broadcastable, Union(None, [1 x #channels],  [(#Rects per batch) x #channels])

File: kornia/utils/draw.py (hunk 5 above)
Comment (edgarriba) [reply kornia/utils/draw.py:17]: @shijianjian why [y,x] instead of [x,y]?

**Comment by edgarriba** [comment]
@anguelos I like the idea of the image batch id. However, I'd suggest to have a separated argument so that the actual location can be in a tensor that in future is purely backpropagable because I don't think we can pass gradients to image indices for now.

@mmathew23 What do you think about this ?

**Comment by shijianjian** [comment]
I kinda agree with the 5-dim tensor representation for the idea of non-fixed number of boxes. The only concern here is that it would be inconsistent with our previous tensor philosophy (starts with batch size). If we are implementing this approach, other functions might be needed to be updated as well (e.g. cropping, perspective) to be more consistent.

This might be another PR to do.

**Comment by edgarriba** [comment]
I agree with Jian, that why I suggested also to have it as a separated argument too. If remember well in some operators in torchvision (nms?) they use this same practice. We should try to align with them in that sense, also with bbox representation.

**Comment by mmathew23** [comment]
@edgarriba I don't follow how separating the argument would allow for backdproping if pytorch supported passing gradients to image indices.  

I do, however, like the idea to use image batch id so that the number of rectangles created can be flexible. I guess the only question is whether or not to separate the batch index argument.

**Comment by edgarriba** [comment]
@mmathew23 we have been looking at torchvision [`draw_bounding_box`](https://github.com/oke-aditya/vision/blob/master/torchvision/utils.py#L139) with @lferraz and I think we could follow a similar convention.

I propose the following API:

```python
def draw_rectangle(
    image: torch.Tensor,
    rectangle: torch.Tensor,
    color: Optional[Union[float, torch.Tensor]] = 1.0,
    fill: Optional[bool] = False,
    width: int = 1,
    batch_id: Optional[List[int]] = None
) -> torch.Tensor:
"""Draws N rectangles on a batch of image tensors
    Args:
        image (torch.Tensor): is tensor of BxCxHxW
        rectangle (torch.Tensor): represents number of rectangles to draw in Nx4
            N is the number of boxes to draw per batch [x1, y1, x2, y2]
            4 is in (top_left.x, top_left.y, bot_right.x, bot_right.y)
        color (int or torch.Tensor, optional): can be a float or tensor if image channels > 1
            if float is used and channels > 1, float will be broadcasted to all channels. Default: 1.0.
        fill (bool, optional): is a flag used to fill the boxes with color if True. Default: False.
        width (int): The line width. Default: 1. (Not implemented yet)
        batch_id (List[int], optional): A tensor corresponding to the indices of the image batch for each box.

    Returns:
        torch.Tensor: This operation modifies image inplace but also returns the drawn tensor for convenience with same shape the input BxCxHxW.

    Example:
        >>> img = torch.rand(2, 3, 10, 12)
        >>> rect = torch.tensor([[0, 0, 4, 4], [4, 4, 10, 10]])
        >>> out = draw_rectangle(img, rect, batch_id=[0, 1])
"""
```

**Comment by edgarriba** [comment]
@mmathew23 make we could drop for now Union in color since it will produce some issues with jit. Let me know whether you need help on this since we are preparing the release for pytorch 1.8 for this next week and would be very useful to have this feature in (as experimental)

**Comment by mmathew23** [comment]
@edgarriba I can work on this. If the batch id is not given should we make sure the number of rectangles, N, is the same as B?

**Comment by edgarriba** [comment]
In that case I'd do img (BCHW) rect (BN4) where each image of the batch received N rectangles

**Comment by mmathew23** [comment]
To be clear, are you suggesting to remove the optional batch_id argument and updating rectangle to be BN4, or this is specifically for case where batch_id is not given?

**Comment by mmathew23** [comment]
I went ahead with removing the batch id for now. A test is failing with the error "Slice index must be an integer or None" after I removed the int() calls in favor of .long() on the rectangle, but would .int() be better?

**Comment by edgarriba** [comment]
we can extend later with the batch_id thing

**Comment by mmathew23** [comment]
@edgarriba not sure why some of these tests are failing. mypy passes locally for me.

**Review by edgarriba — APPROVED** [review]
LGTM. I fixed the mypy issues and added to the docs

**Comment by edgarriba** [comment]
@mmathew23 awesome contribution :))

See here a small example I tried: https://colab.research.google.com/drive/1me_DxgMvsHIheLh-Pao7rmrsafKO5Lg3?usp=sharing
